metrix/src/parser.rs

```diff
@@ -0,0 +1,176 @@
+#[macro_use]
+extern crate nom;
+
+use std::str;
+use std::fmt;
+
+use nom::{
+    character::is_alphanumeric,
+    character::complete::{digit1 as digit, multispace1 as multispace},
+};
+
+pub struct Expression {
+    left: Box<ExpressionType>,
+    operator: String,
+    right: Box<ExpressionType>,
+}
+
+
+pub enum ExpressionType {
+    OuterExpression(Expression),
+    BaseExpression(BaseExpression),
+}
+
+pub struct BaseExpression {
+    field: String,
+    comparator: String,
+    value: String,
+}
+
+pub fn parse_query_string(input: String) -> Result<ExpressionType, &'static str> {
+    match root_expression(input.into_bytes()) {
+        Ok((_, o)) => Ok(o),
+        Err(_) => Err("Failed to parse query string"),
+    }
+}
+
+impl fmt::Display for ExpressionType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ExpressionType::OuterExpression(outer_expression) => write!(f, "{}", outer_expression),
+            ExpressionType::BaseExpression(base_expression) => write!(f, "{}", base_expression),
+        }
+    }
+}
+
+impl fmt::Display for Expression {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "(")?;
+        write!(f, "{}", self.left)?;
+        write!(f, " ")?;
+        write!(f, "{}", self.operator)?;
+        write!(f, " ")?;
+        write!(f, "{}", self.right)?;
+        write!(f, ")")
+    }
+}
+
+impl fmt::Display for BaseExpression {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "(")?;
+        write!(f, "{}", self.field)?;
+        write!(f, " ")?;
+        write!(f, "{}", self.comparator)?;
+        write!(f, " ")?;
+        write!(f, "{}", self.value)?;
+        write!(f, ")")
+    }
+}
+
+named!(root_expression<ExpressionType>,
```

Comment:
> For new parsers, we recommend using the functions instead of macros, since rustc messages will be much easier to understand. 

https://docs.rs/nom/5.0.1/nom/

Did you come across this deprecation? Or I guess it doesn't really matter since they're both supported...